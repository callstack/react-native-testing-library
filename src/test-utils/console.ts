import { format } from 'util';

export function excludeConsoleMessage(logFn: (...args: unknown[]) => void, excludeMessage: string) {
  return (...args: unknown[]) => {
    const message = format(...args);
    if (message.includes(excludeMessage)) {
      return;
    }

    logFn(...args);
  };
}
