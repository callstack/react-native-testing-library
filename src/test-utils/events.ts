interface EventEntry {
  name: string;
  payload: any;
}

export function createEventLogger() {
  const events: EventEntry[] = [];
  const logEvent = (name: string) => {
    return (event: unknown) => {
      const eventEntry: EventEntry = {
        name,
        payload: event,
      };

      events.push(eventEntry);
    };
  };

  return { events, logEvent };
}
