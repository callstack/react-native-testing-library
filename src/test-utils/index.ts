export * from './console';
export * from './events';
