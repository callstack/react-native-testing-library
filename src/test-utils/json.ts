import type { ReactTestRendererJSON } from 'react-test-renderer';

type JsonPropsMapper = {
  [key: string]: unknown;
};

export function mapJsonProps<T extends ReactTestRendererJSON | ReactTestRendererJSON[] | null>(
  element: T,
  mapper: JsonPropsMapper,
): T {
  if (Array.isArray(element)) {
    return element.map((e) => mapJsonProps(e, mapper)) as T;
  }

  if (!element) {
    return element;
  }

  const resultProps = { ...element.props };
  Object.keys(mapper).forEach((key) => {
    if (key in element.props) {
      resultProps[key] = mapper[key];
    }
  });

  const resultElement = { ...element, props: resultProps };
  Object.defineProperty(resultElement, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return resultElement;
}
