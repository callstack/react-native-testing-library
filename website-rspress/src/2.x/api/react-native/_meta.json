[
  {
    "type": "file",
    "name": "index",
    "label": "Table of Contents"
  },
  {
    "type": "dir",
    "name": "classes",
    "label": "Classes"
  },
  {
    "type": "dir",
    "name": "interfaces",
    "label": "Interfaces"
  },
  {
    "type": "dir",
    "name": "types",
    "label": "Types"
  }
]
