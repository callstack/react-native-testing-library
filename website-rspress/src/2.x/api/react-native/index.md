# Re.Pack React Native APIs

## Table of contents

### Classes

- [Chunk](./classes/Chunk.md)
- [ChunkManager](./classes/ChunkManager.md)

### Interfaces

- [ChunkConfig](./interfaces/ChunkConfig.md)
- [ChunkManagerConfig](./interfaces/ChunkManagerConfig.md)
- [RemoteChunkLocation](./interfaces/RemoteChunkLocation.md)
- [StorageApi](./interfaces/StorageApi.md)

### Type aliases

- [RemoteChunkResolver](./types/RemoteChunkResolver.md)
