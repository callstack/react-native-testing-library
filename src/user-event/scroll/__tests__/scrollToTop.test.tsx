import * as React from 'react';
import { ScrollView, View } from 'react-native';
import { createEventLogger } from '../../../test-utils';
import { render, screen } from '../../..';
import { userEvent } from '../..';

describe('userEvent.scroll with fake timers', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(0);
  });

  test('does NOT work if starting position is x:0, y:0', async () => {
    render(<ScrollView testID="scrollable" />);

    const user = userEvent.setup();
    await expect(
      user.scrollToTop(screen.getByTestId('scrollable'))
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"scrollToTop() does NOT trigger if content offset is already x:0, y:0."`
    );
  });

  test('calls events', async () => {
    const { events, logEvent } = createEventLogger();
    const user = userEvent.setup();

    render(
      <ScrollView
        onScroll={logEvent('scroll')}
        onScrollBeginDrag={logEvent('scrollBeginDrag')}
        onScrollEndDrag={logEvent('scrollEndDrag')}
        onMomentumScrollBegin={logEvent('momentumScrollBegin')}
        onMomentumScrollEnd={logEvent('momentumScrollEnd')}
        onScrollToTop={logEvent('scrollToTop')}
        testID="scrollable"
      />
    );

    await user.scrollTo(screen.getByTestId('scrollable'), { y: 20 });
    await user.scrollToTop(screen.getByTestId('scrollable'));

    expect(events).toMatchSnapshot('scrollTo({ y: 20 }) + scrollToTop()');
  });

  it('does NOT work on View', async () => {
    const screen = render(<View testID="view" />);

    const user = userEvent.setup();
    await expect(
      user.scrollToTop(screen.getByTestId('view'))
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"scrollToTop() works only with host "ScrollView" elements. Passed element has type "View"."`
    );
  });

  test('scrollToTop is accessible directly in userEvent', async () => {
    const mockOnScrollToTop = jest.fn();

    render(
      <ScrollView onScrollToTop={mockOnScrollToTop} testID="scrollable" />
    );

    await userEvent.scrollTo(screen.getByTestId('scrollable'), { y: 20 });
    await userEvent.scrollToTop(screen.getByTestId('scrollable'));
    expect(mockOnScrollToTop).toHaveBeenCalled();
  });
});
