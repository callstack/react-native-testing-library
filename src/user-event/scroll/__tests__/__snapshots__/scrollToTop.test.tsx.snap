// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userEvent.scroll with fake timers calls events: scrollTo({ y: 20 }) + scrollToTop() 1`] = `
[
  {
    "name": "scrollBeginDrag",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 0,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 4,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 8,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 12,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 16,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scrollEndDrag",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 20,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scrollBeginDrag",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 20,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 16,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 12,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 8,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 4,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scrollEndDrag",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 0,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
  {
    "name": "scrollToTop",
    "payload": {
      "contentInset": {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
      "contentOffset": {
        "x": 0,
        "y": 0,
      },
      "contentSize": {
        "height": 0,
        "width": 0,
      },
      "layoutMeasurement": {
        "height": 0,
        "width": 0,
      },
      "responderIgnoreScroll": true,
      "target": 0,
      "velocity": {
        "x": 0,
        "y": 0,
      },
    },
  },
]
`;
